exporter/metric/stdout/stdout.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stdout
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"go.opentelemetry.io/api/core"
+	"go.opentelemetry.io/sdk/export"
+	"go.opentelemetry.io/sdk/metric/aggregator"
+)
+
+type Exporter struct {
+	options Options
+}
+
+// Options are the options to be used when initializing a stdout export.
+type Options struct {
+	// File is the destination.  If not set, os.Stdout is used.
+	File *os.File
+
+	// PrettyPrint will pretty the json representation of the span,
+	// making it print "pretty". Default is false.
+	PrettyPrint bool
+
+	// Quantiles are the desired aggregation quantiles for measure
+	// metric data, used when the configured aggregator supports
+	// quantiles.
+	Quantiles []float64
+}
+
+type expoBatch struct {
+	Timestamp time.Time  `json:"time,omitempty"`
+	Updates   []expoLine `json:"updates,omitempty"`
+}
+
+type expoLine struct {
+	Name      string      `json:"name"`
+	Max       interface{} `json:"max,omitempty"`
+	Sum       interface{} `json:"sum,omitempty"`
+	Count     interface{} `json:"count,omitempty"`
+	LastValue interface{} `json:"last,omitempty"`
+	Timestamp time.Time   `json:"time,omitempty"`
+}
+
+var _ export.MetricExporter = &Exporter{}
+
+func New(options Options) *Exporter {
+	if options.File == nil {
+		options.File = os.Stdout
+	}
+	return &Exporter{
+		options: options,
+	}
+}
+
+func (e *Exporter) Export(_ context.Context, producer export.MetricProducer) {
+	var batch expoBatch
+	producer.Foreach(func(agg export.MetricAggregator, desc *export.Descriptor, labelValues []core.Value) {
```

Comment:
Do we believe that exporters should use parallelism/locking, or be single-threaded?

Reply:
The whole collection pass is single-threaded.  The metric API is lock-free with respect to the collection pass. If there's an interest in speeding the collection pass with the use of multiple goroutines.  The base SDK will be tricky to modify in this regard, the lock-freedom is built on the assumption that the collect pass is single-threaded. However, once the base SDK has finished it's single-threaded pass, the controller could divide work into multiple parts and the exporter could be written with corresponding locking, I suppose, or the exporter could actually use multiple output connections, etc.